generator/validate/gen.go

```diff
@@ -0,0 +1,172 @@
+						if unionDiscriminatorField != nil {
+							root.AddError(loader.ErrFromNode(fmt.Errorf(
+								"Union `%v` should have only 1 union discriminator, but has 2: `%v` and `%v`",
+								typeToCheck.Name,
+								unionDiscriminatorField.Name,
+								unionField.Name), unionField.RawField))
+						}
+						unionDiscriminatorField = &(typeToCheck.Fields[index])
+					} else {
+						unionMembers = append(unionMembers, unionField.Name)
+					}
+				}
+				if unionDiscriminatorField == nil {
+					root.AddError(loader.ErrFromNode(fmt.Errorf(
+						"Union `%v` should have 1 union discriminator. See here for details: https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190325-unions.md#proposal",
+						typeToCheck.Name), typeToCheck.RawSpec))
+				} else {
+					if unionDiscriminatorField.Markers.Get("optional") == nil {
+						root.AddError(loader.ErrFromNode(fmt.Errorf(
+							"In union `%v` the union discriminator `%v` should have the `+optional` comment marker",
+							typeToCheck.Name,
+							unionDiscriminatorField.Name), unionDiscriminatorField.RawField))
+					}
+					if !strings.Contains(unionDiscriminatorField.Tag.Get("json"), ",omitempty") {
+						root.AddError(loader.ErrFromNode(fmt.Errorf(
+							"In union `%v` the union discriminator `%v` should contain the `omitempty` option in its `json` tag, since it is expected to be an optional field",
+							typeToCheck.Name,
+							unionDiscriminatorField.Name), unionDiscriminatorField.RawField))
+					}
+
+					wrongTypeError := loader.ErrFromNode(fmt.Errorf(
+						"In union `%v` the union discriminator `%v` should have a `string` type, or a type (defined in the same package) whose underlying type is a string",
+						typeToCheck.Name,
+						unionDiscriminatorField.Name), unionDiscriminatorField.RawField)
+
+					discriminatorTypeRef, typeFound := root.TypesInfo.Types[unionDiscriminatorField.RawField.Type]
+					if !typeFound {
+						root.AddError(wrongTypeError)
+					}
+
+					underlyingType, underlyingTypeIsBasic := discriminatorTypeRef.Type.Underlying().(*types.Basic)
+					if !underlyingTypeIsBasic || underlyingType.Kind() != types.String {
+						root.AddError(wrongTypeError)
+					}
+
+					getEnumValidation := func(markers markers.MarkerValues, node ast.Node) *crdmarkers.Enum {
+						enumMarkerIf := markers.Get("kubebuilder:validation:Enum")
+						if enumMarkerIf != nil {
+							if enumMarker, isEnum := enumMarkerIf.(crdmarkers.Enum); isEnum {
+								return &enumMarker
+							}
+							root.AddError(loader.ErrFromNode(fmt.Errorf(
+								"In union `%v` the union discriminator `%v` should only have the kubebuilder marker that defines the allowed union values as an Enum (kubebuilder:validation:Enum=....)",
+								typeToCheck.Name,
+								unionDiscriminatorField.Name), node))
+						}
+						return nil
+					}
+
+					enumMarkerPtr := getEnumValidation(unionDiscriminatorField.Markers, unionDiscriminatorField.RawField)
```

Comment:
To add here, I think in the latter case we should force the enum annotation to be defined on the `discriminatorType` rather than in the struct field. I can get past validation by moving the `enum` annotation into the Component struct, for example.

Reply:
The problem is that the checks are interleaved, making it hard to follow:
1. Get enum marker from struct field
2. Check type of struct field; if it's not `string`, then
    * a. Check that type def for struct field's type is in the same package
    * b. Get enum marker for type def
    * c. Check if type def enum marker from b is `nil`; if it is not, then
        * Check whether struct field enum marker from 1. is `nil`; if it is not add an error for two enum markers
        * Otherwise, overwrite enum marker from step 1 with the type def enum marker
3. Check if enum marker is `nil` and add error if it is (not caring if it was set in step 1 or step c.2)
4. Check enum marker's contents to match expected

In the first case (type of struct field is `string`), the check is
1. Get enum marker, ensure not `nil`
2. Check enum marker contents

In the second case (struct field is a named type), the check is (or could be simplified to)
1. Check no enum marker on struct field
2. Check that there is an enum marker on named type
3. Check contents of enum marker

Reply:
Fixed in commit https://github.com/devfile/api/pull/142/commits/5fb3883fc4072194ad8a9a6e1b2f4ee9f5dba691